File: src/main/java/com/aiproxyoauth/model/ModelAliasResolver.java
package com.aiproxyoauth.model;

import java.util.Locale;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class ModelAliasResolver {

    private static final Map<String, ResolvedModel> ALIASES = Map.ofEntries(
            Map.entry("gpt-5.2-codex-low", new ResolvedModel("gpt-5.2-codex", "low")),
            Map.entry("gpt-5.2-codex-medium", new ResolvedModel("gpt-5.2-codex", "medium")),
            Map.entry("gpt-5.2-codex-high", new ResolvedModel("gpt-5.2-codex", "high")),
            Map.entry("gpt-5.2-codex-xhigh", new ResolvedModel("gpt-5.2-codex", "xhigh")),
            Map.entry("gpt-5.1-codex-max-xhigh", new ResolvedModel("gpt-5.1-codex-max", "xhigh")),
            Map.entry("gpt-5.1-none", new ResolvedModel("gpt-5.1", "none"))
    );

    // Junie selects a reasoning tier by sending the model name with a "<base> (<level>)"
    // suffix (it derives the tier menu from the advertised model list). Strip the suffix
    // back into base model + effort; the suffix is the user's explicit tier choice, so it
    // takes precedence over any separately supplied reasoning_effort.
    private static final Pattern REASONING_SUFFIX = Pattern.compile(
            "^(.*?)\\s*\\((low|medium|high|xhigh|minimal|none)\\)$",
            Pattern.CASE_INSENSITIVE
    );

    public record ResolvedModel(String model, String reasoningEffort) {}

    public ResolvedModel resolve(String model) {
        if (model == null) {
            return new ResolvedModel(null, null);
        }
        ResolvedModel resolved = ALIASES.get(model);
        if (resolved != null) {
            return resolved;
        }
        Matcher suffix = REASONING_SUFFIX.matcher(model.trim());
        if (suffix.matches()) {
            return new ResolvedModel(suffix.group(1).trim(), suffix.group(2).toLowerCase(Locale.ROOT));
        }
        return new ResolvedModel(model, null);
    }

    public String clampReasoningEffort(String model, String requestedEffort) {
        if (requestedEffort == null || requestedEffort.isBlank()) {
            return null;
        }

        String effort = requestedEffort.strip().toLowerCase(Locale.ROOT);
        String modelName = model == null ? "" : model.toLowerCase(Locale.ROOT);

        if (isCodexMini(modelName)) {
            return switch (effort) {
                case "high", "xhigh" -> "high";
                default -> "medium";
            };
        }

        if ("minimal".equals(effort)) {
            effort = "low";
        }

        if ("none".equals(effort) && isCodexModel(modelName)) {
            return "low";
        }

        if ("xhigh".equals(effort) && !supportsXHigh(modelName)) {
            return "high";
        }

        return effort;
    }

    private boolean isCodexMini(String modelName) {
        return modelName.contains("codex-mini");
    }

    private boolean isCodexModel(String modelName) {
        return modelName.contains("codex");
    }

    private boolean supportsXHigh(String modelName) {
        // Verified against the live Codex backend: the gpt-5.2/5.3/5.4/5.5 families all
        // accept 'xhigh'. gpt-5.1-codex-max also does; plain gpt-5.1 does not.
        return modelName.equals("gpt-5.1-codex-max")
                || modelName.startsWith("gpt-5.2")
                || modelName.startsWith("gpt-5.3")
                || modelName.startsWith("gpt-5.4")
                || modelName.startsWith("gpt-5.5");
    }
}

File: src/main/kotlin/de/moritzf/proxy/model/ModelAliasResolver.kt
package de.moritzf.proxy.model

import java.util.Locale
import java.util.regex.Pattern

class ModelAliasResolver {
    data class ResolvedModel(
        val model: String?,
        val reasoningEffort: String?,
    ) {
        fun model(): String? = model

        fun reasoningEffort(): String? = reasoningEffort
    }

    fun resolve(model: String?): ResolvedModel {
        if (model == null) {
            return ResolvedModel(null, null)
        }
        val suffix = REASONING_SUFFIX.matcher(model.trim())
        if (suffix.matches()) {
            return ResolvedModel(
                suffix.group(1).trim(),
                suffix.group(2).lowercase(Locale.ROOT),
            )
        }
        return ResolvedModel(model, null)
    }

    fun clampReasoningEffort(model: String?, requestedEffort: String?): String? {
        if (requestedEffort.isNullOrBlank()) {
            return null
        }

        var effort = requestedEffort.trim().lowercase(Locale.ROOT)
        val modelName = model?.lowercase(Locale.ROOT).orEmpty()

        if (isMiniModel(modelName)) {
            // Tested against the live Codex /responses endpoint: mini models accept only
            // medium/high reasoning. Keep the request-time clamp even if unsupported aliases
            // are not advertised, because clients can send stale or manual reasoning_effort values.
            return when (effort) {
                "high", "xhigh" -> "high"
                else -> "medium"
            }
        }

        if (effort == "minimal") {
            effort = "low"
        }

        if (effort == "none" && isCodexModel(modelName)) {
            return "low"
        }

        if (effort == "xhigh" && !supportsXHigh(modelName)) {
            return "high"
        }

        return effort
    }

    private fun isMiniModel(modelName: String): Boolean {
        return modelName.endsWith("-mini") || modelName.contains("codex-mini")
    }

    private fun isCodexModel(modelName: String): Boolean {
        return modelName.contains("codex")
    }

    private fun supportsXHigh(modelName: String): Boolean {
        // Verified against the live Codex backend: the currently supported gpt-5.3/5.4/5.5
        // families all accept 'xhigh'. Older families are no longer reachable on a ChatGPT
        // account, so they are not listed.
        return modelName.startsWith("gpt-5.3") ||
            modelName.startsWith("gpt-5.4") ||
            modelName.startsWith("gpt-5.5")
    }

    companion object {
        // Junie selects a reasoning tier by sending the model name with a "<base> (<level>)"
        // suffix. Strip the suffix back into base model + effort; the suffix is the user's
        // explicit tier choice, so it takes precedence over separate reasoning_effort.
        private val REASONING_SUFFIX = Pattern.compile(
            "^(.*?)\\s*\\((low|medium|high|xhigh|minimal|none)\\)$",
            Pattern.CASE_INSENSITIVE,
        )
    }
}
